src/c++/perf_analyzer/concurrency_worker.cc

```diff
@@ -229,8 +226,15 @@ ConcurrencyWorker::ResetFreeCtxIds()
 {
   std::lock_guard<std::mutex> lock(cb_mtx_);
   free_ctx_ids_ = std::queue<int>();
-  for (size_t i = 0; i < ctxs_.size(); ++i) {
-    free_ctx_ids_.push(i);
+
+  if (ctxs_.size() > 1) {
+    for (size_t i = 0; i < ctxs_.size(); ++i) {
+      free_ctx_ids_.push(i);
+    }
+  } else {
+    for (size_t i = 0; i < thread_config_->concurrency_; i++) {
```

Comment:
old code just had the ctx_id hardcoded to 0 in this case and ignored the free_ctx_ids list. Now it is populated with concurrency_ # of 0's so all of the code can push/pop the same and be used to track concurrency

Reply:
> Now it is populated with concurrency_ # of 0's so all of the code can push/pop the same and be used to track concurrency

I don't understand this. Can you elaborate?

Reply:
I think I understand now, but why is this triggered when contexts already has only one element in it?

Reply:
Oh, I was getting `ctxs_` and `free_ctx_ids_` confused as the same data structure

Reply:
How hard would it be to scrap all of this and just dynamically create InferContext's as the program runs? What's stopping us?

Reply:
It seems unlikely the benefit from (potential) performance improvement from reusing InferContexts outweighs the cost from hindered development by keeping this design.

Reply:
> How hard would it be to scrap all of this and just dynamically create InferContexts as the program runs? What's stopping us?

I think for at least shared memory cases, that would not work, or would be very expensive. It is certainly something we can investigate after we have finished all of the refactoring. Once code is in smaller easy to understand classes it will be much easier for us to change the design.